src/azul/chalice.py

```diff
@@ -145,3 +158,163 @@ def _log_response(self, response):
+        """
+        Returns a mapping of parameter names to their applicable SpecValidator.
+        """
+        http_method = http_method.lower()
+        assert http_method in self.methods
+        return self.validators[http_method]
+
+    def set_route_spec(self, http_method, method_spec):
+        """
+        Sets spec for a give route, creates request parameter SpecValidators
+        """
+        assert http_method.lower() not in self.methods, 'Only specify method_spec once per route path and method'
+        self.specs[http_method] = copy_json(method_spec)
+        self.validators[http_method] = self._create_spec_validators(http_method)
+
+    def extract_parameters(self, http_method: str) -> List[JSON]:
+        """
+        Returns a list of OpenAPI `parameters` that are applicable to a given API route
+        """
+        return [
+            *self.path_spec.get('parameters', []),
+            *self.get_method_spec(http_method).get('parameters', [])
+        ]
+
+    def _create_spec_validators(self, http_method: str) -> Mapping[str, SpecValidator]:
+        spec_validators: Dict[str, SpecValidator] = dict()
+        for parameter in self.extract_parameters(http_method):
+            spec_validator = SpecValidator(**parameter)
+            spec_validators[spec_validator.name] = spec_validator
+        return spec_validators
+
+    def validate_params(self, current_request: Request):
+        def fmt_error(err_description, parameters):
+            joined = ', '.join(f'{p!r}' for p in parameters)
+            return f'{err_description} {pluralize("parameter", len(parameters))} {joined}'
+
+        # OpenAPI requires HTTP method names be lower case
+        http_method = current_request.method.lower()
+
+        query_params = current_request.query_params or {}
+        uri_params = current_request.uri_params or {}
+        all_provided_params = {**query_params, **uri_params}
+
+        provided_params = all_provided_params.keys()
+        mandatory_params = set()
+        validation_params = set()
+
+        spec_parameters = self.get_spec_validators(http_method)
+        for name, spec_parameter in spec_parameters.items():
+            if spec_parameter.required:
+                mandatory_params.add(name)
+                validation_params.add(name)
+            else:
+                validation_params.add(name)
+
+        extra_params = provided_params - (validation_params | mandatory_params)
+        if extra_params:
+            raise BadRequestError(msg=fmt_error('Invalid', extra_params))
+
+        if mandatory_params:
+            missing_params = mandatory_params - provided_params
+            if missing_params:
+                # Sorting is to produce a deterministic error message
+                raise BadRequestError(msg=fmt_error('Missing required',
+                                                    sorted(missing_params)))
+        for param_name in provided_params:
+            if param_name in validation_params:
```

Comment:
My guess is that doing the set operation scrambled the order or the parameters? I just want to confirm that I understand what was going on in the history.

Reply:
Yeah that's correct, the set operation `&=` has messed with the ordering of the variables, I patched this on `develop` as well. https://github.com/DataBiosphere/azul/pull/2359